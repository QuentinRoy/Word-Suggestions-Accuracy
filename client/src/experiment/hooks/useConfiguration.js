import { useRef } from "react";
import { ReadyStates, TaskTypes } from "../../common/constants";
import getTimeZone from "../../common/utils/getTimeZone";
import useJSON from "../../common/hooks/useJson";

const timeZone = getTimeZone();
const timeZoneOffset = new Date().getTimezoneOffset();

const useConfiguration = ({ participant, device, isTest, config }) => {
  const areArgsIncomplete =
    participant == null || device == null || config == null || isTest == null;

  const { current: startDate } = useRef(new Date());
  // In the case there are missing information, providing a null URL to useJson
  // will prevent any loading attempt.
  const [loadingState, baseConfig, error] = useJSON(
    areArgsIncomplete ? null : `./configs/${config}-${device}.json`
  );

  if (areArgsIncomplete) {
    return [ReadyStates.crashed, null, new Error(`Invalid arguments`)];
  }
  if (loadingState === ReadyStates.crashed) {
    return [loadingState, null, error];
  }
  if (loadingState !== ReadyStates.ready) {
    return [loadingState, null, null];
  }
  return [
    loadingState,
    {
      ...baseConfig,
      participant,
      isTest,
      mode: process.env.NODE_ENV,
      gitSha: process.env.REACT_APP_GIT_SHA,
      version: process.env.REACT_APP_VERSION,
      href: window.location.href,
      userAgent: navigator.userAgent,
      isExperimentCompleted: false,
      startDate,
      // This is not much useful since the last experiment is run locally,
      // but I left it here for the sake of consistency.
      timeZone,
      timeZoneOffset,
      [TaskTypes.s3Upload]: {
        filename:
          process.env.NODE_ENV === "development"
            ? `dev/${participant}-${device}-${startDate.toISOString()}.json`
            : `prod/${participant}-${device}-${startDate.toISOString()}.json`,
      },
    },
    null,
  ];
};

export default useConfiguration;
