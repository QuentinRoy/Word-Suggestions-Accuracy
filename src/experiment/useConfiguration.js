import { useMemo } from "react";
import useCorpusFromJson from "./useCorpusFromJson";
import { KeyboardLayouts, LoadingStates } from "../utils/constants";

const defaultAccuracies = [0, 0.25, 0.5, 0.75, 1];
const defaultKeyStrokeDelays = [0, 250, 500, 750, 1000];
const numberOfPracticeTasks = 1;
const numberOfTypingTasks = 2;
const keyboardLayout = KeyboardLayouts.desktop;

const PageArguments = {
  targetAccuracies: "targetAccuracies",
  workerId: "workerId",
  keyStrokeDelays: "keyStrokeDelays"
};

const getPageArguments = () => {
  const urlParams = new URL(document.location).searchParams;
  const workerId = urlParams.get(PageArguments.workerId);
  const keyStrokeDelays = urlParams.has(PageArguments.keyStrokeDelays)
    ? urlParams
        .get(PageArguments.keyStrokeDelays)
        .split(",")
        .map(x => +x)
    : defaultKeyStrokeDelays;
  const targetAccuracies = urlParams.has(PageArguments.targetAccuracies)
    ? urlParams
        .get(PageArguments.targetAccuracies)
        .split(",")
        .map(x => +x)
    : defaultAccuracies;
  const args = {
    participant: workerId,
    keyStrokeDelay:
      keyStrokeDelays[Math.floor(Math.random() * keyStrokeDelays.length)],
    targetAccuracy:
      targetAccuracies[Math.floor(Math.random() * targetAccuracies.length)]
  };
  return args;
};

const { participant, targetAccuracy, keyStrokeDelay } = getPageArguments();

const TypingTask = (id, isPractice, { words, ...distributionInfo }) => ({
  ...distributionInfo,
  task: "TypingTask",
  sksDistribution: words,
  key: id,
  id,
  isPractice
});

const UploadLogS3 = (id, fireAndForget, participantId) => ({
  task: "S3Upload",
  filename: `${participantId}-${new Date().toISOString()}-log.json`,
  key: id,
  fireAndForget
});

const generateTasks = corpus => {
  const tasks = [];

  // Insert practice-related tasks.
  if (numberOfPracticeTasks > 0) {
    tasks.push({
      task: "InformationScreen",
      content: `The ${numberOfPracticeTasks} following tasks are practice tasks`,
      shortcutEnabled: true,
      key: `${tasks.length}`
    });
    for (let i = 0; i < numberOfPracticeTasks; i += 1) {
      tasks.push(UploadLogS3(`${tasks.length}`, true, participant));
      tasks.push(
        TypingTask(
          `${tasks.length}`,
          true,
          corpus.slice(0, numberOfPracticeTasks)[i]
        )
      );
    }
    tasks.push({
      task: "InformationScreen",
      content: "Practice is over, the real experiment begins here",
      key: `${tasks.length}`
    });
  }

  // Insert measured tasks.
  for (let j = 0; j < numberOfTypingTasks; j += 1) {
    tasks.push(UploadLogS3(`${tasks.length}`, true, participant));
    tasks.push(
      TypingTask(
        `${tasks.length}`,
        true,
        corpus.slice(
          numberOfPracticeTasks,
          numberOfPracticeTasks + numberOfTypingTasks
        )[j]
      )
    );
  }

  tasks.push(UploadLogS3(`${tasks.length}`, false, participant));

  return tasks;
};

const useConfiguration = () => {
  const [loadingState, corpus] = useCorpusFromJson(targetAccuracy);
  const config = useMemo(() => {
    if (loadingState === LoadingStates.loaded) {
      return {
        keyStrokeDelay,
        targetAccuracy,
        participant,
        keyboardLayout,
        children: generateTasks(corpus),
        gitSha: process.env.REACT_APP_GIT_SHA,
        version: process.env.REACT_APP_VERSION
      };
    }
    return null;
  }, [loadingState, corpus]);
  return participant == null
    ? [LoadingStates.invalidArguments, null]
    : [loadingState, config];
};

export default useConfiguration;
