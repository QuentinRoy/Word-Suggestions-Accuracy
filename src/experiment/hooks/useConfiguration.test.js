import { generateTasks } from "./useConfiguration";
import fs from "fs-extra";
import path from "path";

const corpus = JSON.parse(
  fs.readFileSync(path.join(__dirname, "./__fixtures__/test-corpus.json"))
).rows;

describe("generateTasks", () => {
  test("create the tasks as expected", () => {
    expect(generateTasks(corpus, "fileName.json")).toMatchSnapshot();
  });
});
