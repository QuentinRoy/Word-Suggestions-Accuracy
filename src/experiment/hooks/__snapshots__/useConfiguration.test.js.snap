// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateTasks create the tasks as expected 1`] = `
Array [
  Object {
    "key": "consent-0",
    "label": "Consent Form",
    "task": "ConsentForm",
  },
  Object {
    "filename": "fileName.json",
    "fireAndForget": true,
    "key": "upload-1",
    "noProgress": true,
    "task": "S3Upload",
  },
  Object {
    "key": "startup-2",
    "label": "Instructions",
    "task": "Startup",
  },
  Object {
    "filename": "fileName.json",
    "fireAndForget": true,
    "key": "upload-3",
    "noProgress": true,
    "task": "S3Upload",
  },
  Object {
    "content": "Now you will complete an interactive tutorial",
    "currentProgress": false,
    "fullProgress": true,
    "key": "info-4",
    "label": "Tutorial",
    "progressLevel": true,
    "shortcutEnabled": true,
    "tasks": Array [
      "InformationScreen",
      "ExperimentProgress",
    ],
  },
  Object {
    "id": "tuto-5",
    "isPractice": true,
    "key": "tuto-5",
    "noProgress": true,
    "task": "Tutorial",
  },
  Object {
    "content": "Continue with the practice tasks",
    "currentProgress": false,
    "fullProgress": true,
    "key": "info-6",
    "label": "Practice",
    "progressLevel": true,
    "shortcutEnabled": true,
    "tasks": Array [
      "InformationScreen",
      "ExperimentProgress",
    ],
  },
  Object {
    "children": Array [
      Object {
        "diffSdWordsKss": 0.0022252747022366293,
        "diffTotalKss": 0.011111111111111072,
        "diffWordsMeanKss": 0.03888888888888886,
        "id": "practice-7-0",
        "isPractice": true,
        "key": "practice-7-0",
        "meanWordsKss": 0.861111111111111,
        "score": 0.0066681929066738505,
        "sdWordsKss": 0.20222527470223664,
        "sksDistribution": Array [
          Object {
            "sks": 2,
            "word": "my ",
          },
          Object {
            "sks": 6,
            "word": "watch ",
          },
          Object {
            "sks": 5,
            "word": "fell ",
          },
          Object {
            "sks": 3,
            "word": "in ",
          },
          Object {
            "sks": 2,
            "word": "the ",
          },
          Object {
            "sks": 6,
            "word": "water ",
          },
        ],
        "task": "TypingTask",
        "totalKss": 0.8888888888888888,
        "usable": true,
      },
      Object {
        "diffSdWordsKss": 0.004040820577345755,
        "diffTotalKss": 1.1102230246251565e-16,
        "diffWordsMeanKss": 0.039999999999999925,
        "id": "practice-7-1",
        "isPractice": true,
        "key": "practice-7-1",
        "meanWordsKss": 0.86,
        "score": 0.002020410288672933,
        "sdWordsKss": 0.19595917942265426,
        "sksDistribution": Array [
          Object {
            "sks": 11,
            "word": "prevailing ",
          },
          Object {
            "sks": 4,
            "word": "wind ",
          },
          Object {
            "sks": 5,
            "word": "from ",
          },
          Object {
            "sks": 2,
            "word": "the ",
          },
          Object {
            "sks": 5,
            "word": "east ",
          },
        ],
        "task": "TypingTask",
        "totalKss": 0.9,
        "usable": true,
      },
      Object {
        "diffSdWordsKss": 0.009956263667129511,
        "diffTotalKss": 0.01176470588235301,
        "diffWordsMeanKss": 0.014285714285714457,
        "id": "practice-7-2",
        "isPractice": true,
        "key": "practice-7-2",
        "meanWordsKss": 0.9142857142857144,
        "score": 0.01086048477474126,
        "sdWordsKss": 0.20995626366712952,
        "sksDistribution": Array [
          Object {
            "sks": 6,
            "word": "never ",
          },
          Object {
            "sks": 4,
            "word": "too ",
          },
          Object {
            "sks": 2,
            "word": "rich ",
          },
          Object {
            "sks": 4,
            "word": "and ",
          },
          Object {
            "sks": 6,
            "word": "never ",
          },
          Object {
            "sks": 4,
            "word": "too ",
          },
          Object {
            "sks": 5,
            "word": "thin ",
          },
        ],
        "task": "TypingTask",
        "totalKss": 0.9117647058823529,
        "usable": true,
      },
    ],
    "currentProgress": true,
    "fullProgress": false,
    "noProgress": true,
    "progressLevel": true,
    "tasks": Array [
      "ExperimentProgress",
    ],
  },
  Object {
    "content": "Practice is over. You may take a break. The real experiment begins immediately after this screen!",
    "currentProgress": false,
    "fullProgress": true,
    "key": "info-8",
    "label": "Experiment",
    "progressLevel": true,
    "shortcutEnabled": true,
    "tasks": Array [
      "InformationScreen",
      "ExperimentProgress",
    ],
  },
  Object {
    "children": Array [
      Object {
        "diffSdWordsKss": 0.06033549900026072,
        "diffTotalKss": 0.013043478260869601,
        "diffWordsMeanKss": 0.0444444444444444,
        "id": "trial-9-0",
        "isPractice": false,
        "key": "trial-9-0",
        "meanWordsKss": 0.8555555555555555,
        "score": 0.03668948863056516,
        "sdWordsKss": 0.1396645009997393,
        "sksDistribution": Array [
          Object {
            "sks": 10,
            "word": "breathing ",
          },
          Object {
            "sks": 2,
            "word": "is ",
          },
          Object {
            "sks": 9,
            "word": "difficult ",
          },
        ],
        "task": "TypingTask",
        "totalKss": 0.9130434782608695,
        "usable": true,
      },
      Object {
        "diffSdWordsKss": 0.006570514175334302,
        "diffTotalKss": 1.1102230246251565e-16,
        "diffWordsMeanKss": 0.01904761904761887,
        "id": "trial-9-1",
        "isPractice": false,
        "key": "trial-9-1",
        "meanWordsKss": 0.880952380952381,
        "score": 0.0032852570876672066,
        "sdWordsKss": 0.1934294858246657,
        "sksDistribution": Array [
          Object {
            "sks": 2,
            "word": "I ",
          },
          Object {
            "sks": 2,
            "word": "can ",
          },
          Object {
            "sks": 4,
            "word": "see ",
          },
          Object {
            "sks": 4,
            "word": "the ",
          },
          Object {
            "sks": 6,
            "word": "rings ",
          },
          Object {
            "sks": 2,
            "word": "on ",
          },
          Object {
            "sks": 7,
            "word": "Saturn ",
          },
        ],
        "task": "TypingTask",
        "totalKss": 0.9,
        "usable": true,
      },
      Object {
        "diffSdWordsKss": 0,
        "diffTotalKss": 0.003225806451612967,
        "diffWordsMeanKss": 0.04999999999999993,
        "id": "trial-9-2",
        "isPractice": false,
        "key": "trial-9-2",
        "meanWordsKss": 0.85,
        "score": 0.0016129032258064835,
        "sdWordsKss": 0.2,
        "sksDistribution": Array [
          Object {
            "sks": 8,
            "word": "physics ",
          },
          Object {
            "sks": 3,
            "word": "and ",
          },
          Object {
            "sks": 10,
            "word": "chemistry ",
          },
          Object {
            "sks": 2,
            "word": "are ",
          },
          Object {
            "sks": 5,
            "word": "hard ",
          },
        ],
        "task": "TypingTask",
        "totalKss": 0.9032258064516129,
        "usable": true,
      },
      Object {
        "diffSdWordsKss": 0.019137867116659518,
        "diffTotalKss": 0.0034482758620688614,
        "diffWordsMeanKss": 0.046666666666666634,
        "id": "trial-9-3",
        "isPractice": false,
        "key": "trial-9-3",
        "meanWordsKss": 0.8533333333333333,
        "score": 0.01129307148936419,
        "sdWordsKss": 0.1808621328833405,
        "sksDistribution": Array [
          Object {
            "sks": 3,
            "word": "my ",
          },
          Object {
            "sks": 3,
            "word": "bank ",
          },
          Object {
            "sks": 8,
            "word": "account ",
          },
          Object {
            "sks": 2,
            "word": "is ",
          },
          Object {
            "sks": 10,
            "word": "overdrawn ",
          },
        ],
        "task": "TypingTask",
        "totalKss": 0.896551724137931,
        "usable": true,
      },
      Object {
        "diffSdWordsKss": 0,
        "diffTotalKss": 0.0034482758620688614,
        "diffWordsMeanKss": 0.04999999999999993,
        "id": "trial-9-4",
        "isPractice": false,
        "key": "trial-9-4",
        "meanWordsKss": 0.85,
        "score": 0.0017241379310344307,
        "sdWordsKss": 0.2,
        "sksDistribution": Array [
          Object {
            "sks": 10,
            "word": "elections ",
          },
          Object {
            "sks": 6,
            "word": "bring ",
          },
          Object {
            "sks": 3,
            "word": "out ",
          },
          Object {
            "sks": 2,
            "word": "the ",
          },
          Object {
            "sks": 5,
            "word": "best ",
          },
        ],
        "task": "TypingTask",
        "totalKss": 0.896551724137931,
        "usable": true,
      },
      Object {
        "diffSdWordsKss": 0.004384341826064231,
        "diffTotalKss": 0.02499999999999991,
        "diffWordsMeanKss": 0.06071428571428561,
        "id": "trial-9-5",
        "isPractice": false,
        "key": "trial-9-5",
        "meanWordsKss": 0.8392857142857143,
        "score": 0.014692170913032071,
        "sdWordsKss": 0.20438434182606424,
        "sksDistribution": Array [
          Object {
            "sks": 3,
            "word": "we ",
          },
          Object {
            "sks": 2,
            "word": "are ",
          },
          Object {
            "sks": 6,
            "word": "having ",
          },
          Object {
            "sks": 10,
            "word": "spaghetti ",
          },
        ],
        "task": "TypingTask",
        "totalKss": 0.875,
        "usable": true,
      },
      Object {
        "diffSdWordsKss": 0.02679491924311228,
        "diffTotalKss": 1.1102230246251565e-16,
        "diffWordsMeanKss": 1.1102230246251565e-16,
        "id": "trial-9-6",
        "isPractice": false,
        "key": "trial-9-6",
        "meanWordsKss": 0.9,
        "score": 0.013397459621556196,
        "sdWordsKss": 0.17320508075688773,
        "sksDistribution": Array [
          Object {
            "sks": 3,
            "word": "time ",
          },
          Object {
            "sks": 3,
            "word": "to ",
          },
          Object {
            "sks": 3,
            "word": "go ",
          },
          Object {
            "sks": 9,
            "word": "shopping ",
          },
        ],
        "task": "TypingTask",
        "totalKss": 0.9,
        "usable": true,
      },
      Object {
        "diffSdWordsKss": 0.004040820577345755,
        "diffTotalKss": 0.01538461538461533,
        "diffWordsMeanKss": 0.039999999999999925,
        "id": "trial-9-7",
        "isPractice": false,
        "key": "trial-9-7",
        "meanWordsKss": 0.86,
        "score": 0.009712717980980543,
        "sdWordsKss": 0.19595917942265426,
        "sksDistribution": Array [
          Object {
            "sks": 2,
            "word": "a ",
          },
          Object {
            "sks": 8,
            "word": "problem ",
          },
          Object {
            "sks": 4,
            "word": "with ",
          },
          Object {
            "sks": 2,
            "word": "the ",
          },
          Object {
            "sks": 7,
            "word": "engine ",
          },
        ],
        "task": "TypingTask",
        "totalKss": 0.8846153846153846,
        "usable": true,
      },
      Object {
        "diffSdWordsKss": 0.004040820577345755,
        "diffTotalKss": 0.0034482758620688614,
        "diffWordsMeanKss": 0.039999999999999925,
        "id": "trial-9-8",
        "isPractice": false,
        "key": "trial-9-8",
        "meanWordsKss": 0.86,
        "score": 0.0037445482197073082,
        "sdWordsKss": 0.19595917942265426,
        "sksDistribution": Array [
          Object {
            "sks": 10,
            "word": "elephants ",
          },
          Object {
            "sks": 2,
            "word": "are ",
          },
          Object {
            "sks": 7,
            "word": "afraid ",
          },
          Object {
            "sks": 3,
            "word": "of ",
          },
          Object {
            "sks": 4,
            "word": "mice ",
          },
        ],
        "task": "TypingTask",
        "totalKss": 0.896551724137931,
        "usable": true,
      },
      Object {
        "diffSdWordsKss": 0,
        "diffTotalKss": 0.011111111111111072,
        "diffWordsMeanKss": 1.1102230246251565e-16,
        "id": "trial-9-9",
        "isPractice": false,
        "key": "trial-9-9",
        "meanWordsKss": 0.9,
        "score": 0.005555555555555536,
        "sdWordsKss": 0.2,
        "sksDistribution": Array [
          Object {
            "sks": 3,
            "word": "my ",
          },
          Object {
            "sks": 9,
            "word": "favorite ",
          },
          Object {
            "sks": 6,
            "word": "place ",
          },
          Object {
            "sks": 3,
            "word": "to ",
          },
          Object {
            "sks": 3,
            "word": "visit ",
          },
        ],
        "task": "TypingTask",
        "totalKss": 0.8888888888888888,
        "usable": true,
      },
      Object {
        "diffSdWordsKss": 0.009059346043506666,
        "diffTotalKss": 0.0034482758620688614,
        "diffWordsMeanKss": 0.02499999999999991,
        "id": "trial-9-10",
        "isPractice": false,
        "key": "trial-9-10",
        "meanWordsKss": 0.875,
        "score": 0.006253810952787764,
        "sdWordsKss": 0.19094065395649334,
        "sksDistribution": Array [
          Object {
            "sks": 6,
            "word": "three ",
          },
          Object {
            "sks": 2,
            "word": "two ",
          },
          Object {
            "sks": 3,
            "word": "one ",
          },
          Object {
            "sks": 5,
            "word": "zero ",
          },
          Object {
            "sks": 6,
            "word": "blast ",
          },
          Object {
            "sks": 4,
            "word": "off ",
          },
        ],
        "task": "TypingTask",
        "totalKss": 0.896551724137931,
        "usable": true,
      },
      Object {
        "diffSdWordsKss": 0,
        "diffTotalKss": 0.01764705882352935,
        "diffWordsMeanKss": 1.1102230246251565e-16,
        "id": "trial-9-11",
        "isPractice": false,
        "key": "trial-9-11",
        "meanWordsKss": 0.9,
        "score": 0.008823529411764675,
        "sdWordsKss": 0.2,
        "sksDistribution": Array [
          Object {
            "sks": 3,
            "word": "my ",
          },
          Object {
            "sks": 9,
            "word": "favorite ",
          },
          Object {
            "sks": 4,
            "word": "subject ",
          },
          Object {
            "sks": 3,
            "word": "is ",
          },
          Object {
            "sks": 11,
            "word": "psychology ",
          },
        ],
        "task": "TypingTask",
        "totalKss": 0.8823529411764706,
        "usable": true,
      },
      Object {
        "diffSdWordsKss": 0.019826002557464695,
        "diffTotalKss": 0.003225806451612967,
        "diffWordsMeanKss": 0.09230769230769209,
        "id": "trial-9-12",
        "isPractice": false,
        "key": "trial-9-12",
        "meanWordsKss": 0.8076923076923078,
        "score": 0.01152590450453883,
        "sdWordsKss": 0.2198260025574647,
        "sksDistribution": Array [
          Object {
            "sks": 14,
            "word": "circumstances ",
          },
          Object {
            "sks": 2,
            "word": "are ",
          },
          Object {
            "sks": 12,
            "word": "unacceptable ",
          },
        ],
        "task": "TypingTask",
        "totalKss": 0.9032258064516129,
        "usable": true,
      },
      Object {
        "diffSdWordsKss": 0.009059346043506666,
        "diffTotalKss": 0.00909090909090915,
        "diffWordsMeanKss": 0.02499999999999991,
        "id": "trial-9-13",
        "isPractice": false,
        "key": "trial-9-13",
        "meanWordsKss": 0.875,
        "score": 0.009075127567207908,
        "sdWordsKss": 0.19094065395649334,
        "sksDistribution": Array [
          Object {
            "sks": 6,
            "word": "watch ",
          },
          Object {
            "sks": 2,
            "word": "out ",
          },
          Object {
            "sks": 3,
            "word": "for ",
          },
          Object {
            "sks": 4,
            "word": "low ",
          },
          Object {
            "sks": 7,
            "word": "flying ",
          },
          Object {
            "sks": 8,
            "word": "objects ",
          },
        ],
        "task": "TypingTask",
        "totalKss": 0.9090909090909091,
        "usable": true,
      },
      Object {
        "diffSdWordsKss": 0.006570514175334302,
        "diffTotalKss": 0.003225806451612967,
        "diffWordsMeanKss": 0.01904761904761887,
        "id": "trial-9-14",
        "isPractice": false,
        "key": "trial-9-14",
        "meanWordsKss": 0.880952380952381,
        "score": 0.0048981603134736346,
        "sdWordsKss": 0.1934294858246657,
        "sksDistribution": Array [
          Object {
            "sks": 3,
            "word": "if ",
          },
          Object {
            "sks": 3,
            "word": "at ",
          },
          Object {
            "sks": 6,
            "word": "first ",
          },
          Object {
            "sks": 4,
            "word": "you ",
          },
          Object {
            "sks": 2,
            "word": "do ",
          },
          Object {
            "sks": 2,
            "word": "not ",
          },
          Object {
            "sks": 8,
            "word": "succeed ",
          },
        ],
        "task": "TypingTask",
        "totalKss": 0.9032258064516129,
        "usable": true,
      },
      Object {
        "diffSdWordsKss": 0.0022252747022366015,
        "diffTotalKss": 0.01764705882352935,
        "diffWordsMeanKss": 0.03888888888888875,
        "id": "trial-9-15",
        "isPractice": false,
        "key": "trial-9-15",
        "meanWordsKss": 0.8611111111111112,
        "score": 0.009936166762882975,
        "sdWordsKss": 0.2022252747022366,
        "sksDistribution": Array [
          Object {
            "sks": 7,
            "word": "please ",
          },
          Object {
            "sks": 8,
            "word": "provide ",
          },
          Object {
            "sks": 5,
            "word": "your ",
          },
          Object {
            "sks": 5,
            "word": "date ",
          },
          Object {
            "sks": 2,
            "word": "of ",
          },
          Object {
            "sks": 3,
            "word": "birth ",
          },
        ],
        "task": "TypingTask",
        "totalKss": 0.8823529411764706,
        "usable": true,
      },
      Object {
        "diffSdWordsKss": 0.0022252747022366015,
        "diffTotalKss": 0.011111111111111072,
        "diffWordsMeanKss": 0.03888888888888875,
        "id": "trial-9-16",
        "isPractice": false,
        "key": "trial-9-16",
        "meanWordsKss": 0.8611111111111112,
        "score": 0.006668192906673837,
        "sdWordsKss": 0.2022252747022366,
        "sksDistribution": Array [
          Object {
            "sks": 3,
            "word": "we ",
          },
          Object {
            "sks": 2,
            "word": "run ",
          },
          Object {
            "sks": 4,
            "word": "the ",
          },
          Object {
            "sks": 5,
            "word": "risk ",
          },
          Object {
            "sks": 2,
            "word": "of ",
          },
          Object {
            "sks": 8,
            "word": "failure ",
          },
        ],
        "task": "TypingTask",
        "totalKss": 0.8888888888888888,
        "usable": true,
      },
      Object {
        "diffSdWordsKss": 0.019137867116659518,
        "diffTotalKss": 0.003225806451612967,
        "diffWordsMeanKss": 0.046666666666666634,
        "id": "trial-9-17",
        "isPractice": false,
        "key": "trial-9-17",
        "meanWordsKss": 0.8533333333333333,
        "score": 0.011181836784136243,
        "sdWordsKss": 0.1808621328833405,
        "sksDistribution": Array [
          Object {
            "sks": 7,
            "word": "prayer ",
          },
          Object {
            "sks": 2,
            "word": "in ",
          },
          Object {
            "sks": 8,
            "word": "schools ",
          },
          Object {
            "sks": 8,
            "word": "offends ",
          },
          Object {
            "sks": 3,
            "word": "some ",
          },
        ],
        "task": "TypingTask",
        "totalKss": 0.9032258064516129,
        "usable": true,
      },
      Object {
        "diffSdWordsKss": 0.023606797749978953,
        "diffTotalKss": 1.1102230246251565e-16,
        "diffWordsMeanKss": 1.1102230246251565e-16,
        "id": "trial-9-18",
        "isPractice": false,
        "key": "trial-9-18",
        "meanWordsKss": 0.9,
        "score": 0.011803398874989532,
        "sdWordsKss": 0.22360679774997896,
        "sksDistribution": Array [
          Object {
            "sks": 3,
            "word": "he ",
          },
          Object {
            "sks": 3,
            "word": "is ",
          },
          Object {
            "sks": 5,
            "word": "just ",
          },
          Object {
            "sks": 2,
            "word": "like ",
          },
          Object {
            "sks": 9,
            "word": "everyone ",
          },
          Object {
            "sks": 5,
            "word": "else ",
          },
        ],
        "task": "TypingTask",
        "totalKss": 0.9,
        "usable": true,
      },
      Object {
        "diffSdWordsKss": 0,
        "diffTotalKss": 0.003225806451612967,
        "diffWordsMeanKss": 1.1102230246251565e-16,
        "id": "trial-9-19",
        "isPractice": false,
        "key": "trial-9-19",
        "meanWordsKss": 0.9,
        "score": 0.0016129032258064835,
        "sdWordsKss": 0.2,
        "sksDistribution": Array [
          Object {
            "sks": 6,
            "word": "great ",
          },
          Object {
            "sks": 12,
            "word": "disturbance ",
          },
          Object {
            "sks": 3,
            "word": "in ",
          },
          Object {
            "sks": 4,
            "word": "the ",
          },
          Object {
            "sks": 3,
            "word": "force ",
          },
        ],
        "task": "TypingTask",
        "totalKss": 0.9032258064516129,
        "usable": true,
      },
    ],
    "currentProgress": true,
    "fullProgress": false,
    "noProgress": true,
    "progressLevel": true,
    "tasks": Array [
      "ExperimentProgress",
    ],
  },
  Object {
    "filename": "fileName.json",
    "fireAndForget": true,
    "key": "upload-10",
    "noProgress": true,
    "task": "S3Upload",
  },
  Object {
    "content": "Please fill in the experiment questionnaire",
    "currentProgress": false,
    "fullProgress": true,
    "key": "info-11",
    "label": "Questionnaire",
    "progressLevel": true,
    "shortcutEnabled": true,
    "tasks": Array [
      "InformationScreen",
      "ExperimentProgress",
    ],
  },
  Object {
    "key": "12",
    "noProgress": true,
    "task": "EndQuestionnaire",
  },
  Object {
    "filename": "fileName.json",
    "fireAndForget": true,
    "key": "upload-13",
    "noProgress": true,
    "task": "S3Upload",
  },
  Object {
    "content": "To finish, please complete a few additional typing tasks without impairment or suggestions.",
    "currentProgress": false,
    "fullProgress": true,
    "key": "info-14",
    "label": "Typing speed",
    "progressLevel": true,
    "shortcutEnabled": true,
    "tasks": Array [
      "InformationScreen",
      "ExperimentProgress",
    ],
  },
  Object {
    "children": Array [
      Object {
        "diffSdWordsKss": 0.02679491924311228,
        "diffTotalKss": 0.013043478260869601,
        "diffWordsMeanKss": 1.1102230246251565e-16,
        "id": "typing-15-0",
        "isPractice": false,
        "key": "typing-15-0",
        "meanWordsKss": 0.9,
        "score": 0.01991919875199094,
        "sdWordsKss": 0.17320508075688773,
        "sksDistribution": Array [
          Object {
            "sks": 5,
            "word": "love ",
          },
          Object {
            "sks": 6,
            "word": "means ",
          },
          Object {
            "sks": 3,
            "word": "many ",
          },
          Object {
            "sks": 7,
            "word": "things ",
          },
        ],
        "task": "TypingTask",
        "totalKss": 0.9130434782608695,
        "usable": true,
      },
      Object {
        "diffSdWordsKss": 0,
        "diffTotalKss": 0.01666666666666672,
        "diffWordsMeanKss": 1.1102230246251565e-16,
        "id": "typing-15-1",
        "isPractice": false,
        "key": "typing-15-1",
        "meanWordsKss": 0.9,
        "score": 0.00833333333333336,
        "sdWordsKss": 0.2,
        "sksDistribution": Array [
          Object {
            "sks": 4,
            "word": "you ",
          },
          Object {
            "sks": 5,
            "word": "must ",
          },
          Object {
            "sks": 3,
            "word": "be ",
          },
          Object {
            "sks": 8,
            "word": "getting ",
          },
          Object {
            "sks": 2,
            "word": "old ",
          },
        ],
        "task": "TypingTask",
        "totalKss": 0.9166666666666666,
        "usable": true,
      },
      Object {
        "diffSdWordsKss": 2.7755575615628914e-17,
        "diffTotalKss": 0.004761904761904856,
        "diffWordsMeanKss": 1.1102230246251565e-16,
        "id": "typing-15-2",
        "isPractice": false,
        "key": "typing-15-2",
        "meanWordsKss": 0.9,
        "score": 0.002380952380952442,
        "sdWordsKss": 0.20000000000000004,
        "sksDistribution": Array [
          Object {
            "sks": 2,
            "word": "the ",
          },
          Object {
            "sks": 6,
            "word": "world ",
          },
          Object {
            "sks": 3,
            "word": "is ",
          },
          Object {
            "sks": 2,
            "word": "a ",
          },
          Object {
            "sks": 6,
            "word": "stage ",
          },
        ],
        "task": "TypingTask",
        "totalKss": 0.9047619047619048,
        "usable": true,
      },
      Object {
        "diffSdWordsKss": 0.013661001875017509,
        "diffTotalKss": 1.1102230246251565e-16,
        "diffWordsMeanKss": 0.016666666666666607,
        "id": "typing-15-3",
        "isPractice": false,
        "key": "typing-15-3",
        "meanWordsKss": 0.8833333333333333,
        "score": 0.00683050093750881,
        "sdWordsKss": 0.1863389981249825,
        "sksDistribution": Array [
          Object {
            "sks": 2,
            "word": "can ",
          },
          Object {
            "sks": 2,
            "word": "I ",
          },
          Object {
            "sks": 6,
            "word": "skate ",
          },
          Object {
            "sks": 4,
            "word": "with ",
          },
          Object {
            "sks": 7,
            "word": "sister ",
          },
          Object {
            "sks": 6,
            "word": "today ",
          },
        ],
        "task": "TypingTask",
        "totalKss": 0.9,
        "usable": true,
      },
      Object {
        "diffSdWordsKss": 0.000041653650352801996,
        "diffTotalKss": 0.014285714285714235,
        "diffWordsMeanKss": 0.01836734693877562,
        "id": "typing-15-4",
        "isPractice": false,
        "key": "typing-15-4",
        "meanWordsKss": 0.9183673469387755,
        "score": 0.0071636839680335185,
        "sdWordsKss": 0.1999583463496472,
        "sksDistribution": Array [
          Object {
            "sks": 8,
            "word": "neither ",
          },
          Object {
            "sks": 2,
            "word": "a ",
          },
          Object {
            "sks": 9,
            "word": "borrower ",
          },
          Object {
            "sks": 4,
            "word": "nor ",
          },
          Object {
            "sks": 2,
            "word": "a ",
          },
          Object {
            "sks": 3,
            "word": "lender ",
          },
          Object {
            "sks": 3,
            "word": "be ",
          },
        ],
        "task": "TypingTask",
        "totalKss": 0.8857142857142857,
        "usable": true,
      },
    ],
    "currentProgress": true,
    "fullProgress": false,
    "keyStrokeDelay": 0,
    "noProgress": true,
    "progressLevel": true,
    "suggestionsType": "NONE",
    "tasks": Array [
      "ExperimentProgress",
    ],
  },
  Object {
    "key": "feedbacks-16",
    "task": "FinalFeedbacks",
  },
  Object {
    "key": "inject-end-17",
    "noProgress": true,
    "task": "InjectEnd",
  },
  Object {
    "filename": "fileName.json",
    "fireAndForget": false,
    "key": "upload-18",
    "noProgress": true,
    "task": "S3Upload",
  },
  Object {
    "key": "end-19",
    "task": "EndExperiment",
  },
]
`;
